internal/model/tempnameutil.go

```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"crypto/md5"
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+type tempNamer struct {
+	prefix string
+}
+
+func (t tempNamer) IsTemporary(name string) bool {
+	return strings.HasPrefix(filepath.Base(name), t.prefix)
+}
+
+func (t tempNamer) TempName(name string) string {
+	tdir := filepath.Dir(name)
+	tbase := filepath.Base(name)
+	if len(tbase) > 100 {
+		hash := md5.New()
+		hash.Write([]byte(name))
+		tbase = fmt.Sprintf("%s%s", tbase[0:100], hash.Sum(nil))
```

Comment:
%s here for hash.Sum ought to be random binary crap which might not be acceptable? Probably want %x?